File: dws-step/src/main/java/io/dws/step/config/SingleNodeDefinitionLoader.java
package io.dws.step.config;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.regex.Pattern;

/** Loads and validates the Step half of the shared single-node definition contract. */
public class SingleNodeDefinitionLoader {

  static final String DEFINITION_PATH_ENV = "DWS_STEP_DEFINITION_PATH";
  private static final Pattern DNS_1123_LABEL = Pattern.compile("[a-z0-9]([-a-z0-9]*[a-z0-9])?");

  private final ObjectMapper mapper;
  private final String definitionPath;

  public SingleNodeDefinitionLoader(ObjectMapper mapper, String definitionPath) {
    this.mapper = mapper;
    this.definitionPath = definitionPath;
  }

  public SingleNodeDefinition load() {
    if (definitionPath == null || definitionPath.isBlank()) {
      throw new DefinitionLoadException(DEFINITION_PATH_ENV + " is required but was not set");
    }

    JsonNode definition = readDefinition();
    if (definition == null || !definition.isObject()) {
      throw new DefinitionLoadException("single-node definition must be a JSON object");
    }

    String workflow = requiredText(definition, "workflow");
    String version = requiredText(definition, "version");
    String nodeId = requiredText(definition, "nodeId");
    if (nodeId.length() > 63 || !DNS_1123_LABEL.matcher(nodeId).matches()) {
      throw new DefinitionLoadException("nodeId must be a DNS-1123 label: '" + nodeId + "'");
    }
    if (!"step".equals(requiredText(definition, "kind"))) {
      throw new DefinitionLoadException("definition kind must be 'step'");
    }

    JsonNode task = definition.path("task");
    if (!task.isObject() || task.size() == 0) {
      throw new DefinitionLoadException("step definition must contain a non-empty object 'task'");
    }

    boolean delegatesToFunction = task.has("call") || task.has("run");
    JsonNode functionAppId = definition.get("functionAppId");
    if (delegatesToFunction
        && (functionAppId == null
            || !functionAppId.isTextual()
            || functionAppId.textValue().isBlank())) {
      throw new DefinitionLoadException(
          "functionAppId is required when task.call or task.run is set");
    }
    if (!delegatesToFunction && functionAppId != null) {
      throw new DefinitionLoadException(
          "functionAppId must be absent unless task.call or task.run is set");
    }

    return new SingleNodeDefinition(
        workflow, version, nodeId, task, functionAppId == null ? null : functionAppId.textValue());
  }

  private JsonNode readDefinition() {
    try {
      return mapper.readTree(Files.readString(Path.of(definitionPath)));
    } catch (IOException | IllegalArgumentException e) {
      throw new DefinitionLoadException(
          "failed to load definition '" + definitionPath + "': " + e.getMessage(), e);
    }
  }

  private String requiredText(JsonNode definition, String field) {
    JsonNode value = definition.get(field);
    if (value == null || !value.isTextual() || value.textValue().isBlank()) {
      throw new DefinitionLoadException("definition must contain non-empty string '" + field + "'");
    }
    return value.textValue();
  }
}

File: dws-flow/src/SingleNodeDefinitionLoader.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace Dws.Flow;

/// <summary>Loads and validates the Flow half of the shared single-node definition contract.</summary>
public sealed partial class SingleNodeDefinitionLoader
{
    public const string DefinitionPathEnvironmentVariable = "DWS_FLOW_DEFINITION_PATH";

    private static readonly HashSet<string> ValidScopes = ["main", "for", "try", "catch", "forkBranch"];
    private readonly string? definitionPath;

    public SingleNodeDefinitionLoader(string? definitionPath)
    {
        this.definitionPath = definitionPath;
    }

    public SingleNodeDefinition Load()
    {
        if (string.IsNullOrWhiteSpace(definitionPath))
        {
            throw new DefinitionLoadException($"{DefinitionPathEnvironmentVariable} is required but was not set");
        }

        JsonObject definition = ReadDefinition();
        string workflow = RequiredString(definition, "workflow");
        string version = RequiredString(definition, "version");
        string nodeId = RequiredString(definition, "nodeId");
        if (nodeId.Length > 63 || !Dns1123Label().IsMatch(nodeId))
        {
            throw new DefinitionLoadException($"nodeId must be a DNS-1123 label: '{nodeId}'");
        }

        if (RequiredString(definition, "kind") != "flow")
        {
            throw new DefinitionLoadException("definition kind must be 'flow'");
        }

        string scope = RequiredString(definition, "scope");
        if (!ValidScopes.Contains(scope))
        {
            throw new DefinitionLoadException($"scope must be one of: {string.Join(", ", ValidScopes)}");
        }

        if (definition["tasks"] is not JsonArray tasks || tasks.Any(task => task is not JsonObject taskObject || taskObject.Count == 0))
        {
            throw new DefinitionLoadException("flow definition must contain array 'tasks' of non-empty task objects");
        }

        if (definition["children"] is not JsonObject children || !ChildrenAreValid(children))
        {
            throw new DefinitionLoadException("flow definition must contain object 'children' with non-empty app ID values");
        }

        string? catchAppId = OptionalString(definition, "catch");
        if (definition.ContainsKey("catch") && string.IsNullOrWhiteSpace(catchAppId))
        {
            throw new DefinitionLoadException("catch must be a non-empty app ID string when present");
        }

        return new SingleNodeDefinition(workflow, version, nodeId, scope, tasks, children, catchAppId);
    }

    private JsonObject ReadDefinition()
    {
        try
        {
            return JsonNode.Parse(File.ReadAllText(definitionPath!)) as JsonObject
                ?? throw new DefinitionLoadException("single-node definition must be a JSON object");
        }
        catch (DefinitionLoadException)
        {
            throw;
        }
        catch (Exception exception) when (exception is IOException or JsonException or ArgumentException)
        {
            throw new DefinitionLoadException($"failed to load definition '{definitionPath}': {exception.Message}", exception);
        }
    }

    private static string RequiredString(JsonObject definition, string field)
    {
        if (definition[field] is JsonValue value && value.TryGetValue<string>(out string? text) && !string.IsNullOrWhiteSpace(text))
        {
            return text;
        }

        throw new DefinitionLoadException($"definition must contain non-empty string '{field}'");
    }

    private static string? OptionalString(JsonObject definition, string field)
    {
        return definition[field] is JsonValue value && value.TryGetValue<string>(out string? text) ? text : null;
    }

    private static bool ChildrenAreValid(JsonObject children)
    {
        foreach ((string _, JsonNode? value) in children)
        {
            if (value is not JsonValue jsonValue || !jsonValue.TryGetValue<string>(out string? appId) || string.IsNullOrWhiteSpace(appId))
            {
                return false;
            }
        }

        return true;
    }

    [GeneratedRegex("^[a-z0-9]([-a-z0-9]*[a-z0-9])?$")]
    private static partial Regex Dns1123Label();
}
